### PR TITLE
enhancement/Issue-1118: Single File Bundles for SSR and API routes

### LINKED ISSUE #1118: single file bundles (SFBs) for SSR page and API routes
## Summary
Coming out of the PR for #1088 wanted to track an issue around the output of the bundle files created for SSR page and API route bundles.  It would be nice to make sure all these files are as self contained as possible to not have to rely on any particular bundling setups for different hosting providers.

In particular this applies to SSR pages, as currently two output files are generated
- `__<page>.js` - the entry point file used to invoke a `_<page>.js` file that exports itself as a `handler` function (like in the `server` command)
- `_<page>.js` - the bundled output of `executeModuleUrl`
- As well as _execute-route-module.js_. 

Another (potential) complication / urgency for this has been observed in https://github.com/thescientist13/greenwood-demo-adapter-vercel-lit/issues/3 where in now our page output bundling is not consistent, because we have to do hacky things [like this](https://github.com/ProjectEvergreen/greenwood/blob/v0.29.0-alpha.5/packages/cli/src/config/rollup.config.js#L277).  😬 

In general there should be a way to collapse down a lot of these files into one.  🙏 

## Details
One particular callout on this is that **WCC** take a `URL` for `renderToString` as the entry point, which means at least one file has to exist on disk I guess, meaning there will always be an entry point and then the actual userland code?

In addition this would also mean making sure Rollup doesn't create chunks.

edit: also, coming out #1122 and the actual prototypes for the Vercel and Netlify adapters, the current implementation of adapter plugins by their current design _also_ introduce an entry point, which also seems by nature of the design, e.g.
```sh
functions/
  my-function/
    my-function.js // userland entrypoint
    index.js // adapter wrapper that calls my-function.js
    package.json
```

Also see examples of this in the adapter plugins that have to copy a bunch of files around when preparing for uploading to the respective hosting platform
- #1128 

----

> _**Note:** Assuming we get through #1008 without any issues like if things break _without_ only a single file, otherwise as this is more of an enhancement / optimization, this could either be a 1.0 or post 1.0 feature.  Tracking in the 1.0 Milestone for now ._

### PR DESCRIPTION
**Pull request by DevLab2425**
## Related Issue
Resolves #1118 

## Summary of Changes
- Refactor the API Rollup config generation to create a single entry point _for each API route_, rather than a single Rollup process _for all API routes_. 
- Refactor the SSR Rollup config generation to mimic the API generation, in that, the SSR array is now separate input config files.
- Refactor SSR pages entry point "facade" implementation in bundle lifecycle to use `URL` so we can bundle instead of the hacky string replace solution
- Documented breaking changes in adapter plugin docs

> Demo implementation in - https://github.com/ProjectEvergreen/greenwood-demo-adapter-vercel/pull/22

## TODO
1. [x] Documentation
1. [x] automatic SSR chunks / bundle mapping (nice to have)
    - actually, this is just the same SSR pages issue with us now using `URL`, otherwise it would just be inlined, fwiw
    - just did this for API routes for now, will circle back on SSR pages as part of https://github.com/ProjectEvergreen/greenwood/issues/1163
1. [x] fix windows test cases
1. [x] WCC Upstreams and delete _patches/_ + `postinstall`
    - Some sort of bug with clobbered custom element definitions -  https://github.com/ProjectEvergreen/wcc/issues/144
    - Seems like I may have introduced an infinite loop bug? 😮‍💨  - https://github.com/ProjectEvergreen/wcc/issues/147
1. [x] Should create a discussion to see if there's a better alternative to the URL mechanism, since its good for bundling, but forces a chunk since it really can't be inlined... which isn't the worst thing in the world I suppose 🤷‍♂️ 
    - should capture this a comment in adapter plugins that have to find / filter for route chunk files to copy over to adapter output directories
    - https://github.com/ProjectEvergreen/greenwood/discussions/1204

### CODE REVIEW

**Comment by DevLab2425** [comment]
@thescientist13 

I'm not totally sure if this PR is ready. At the moment, there are 50+ test failures, all seemingly around the fetch call to the Artists API. One example is here: https://github.com/ProjectEvergreen/greenwood/blob/8fad6f78d337c1d836abc1571e5ba34590c41487/packages/cli/test/cases/build.default.ssr-static-export/src/pages/artists.js#L34

If that call does not return an array, the suite fails. I attempted to add some defensive code as a test, which resulted in a single failure of a spec due to a mismatch of values. The code below allows for the process to continue, but it doesn't produce the desired outcome...obviously.

```js
const artistsListItems = Array.isArray(artists) ? artists : []
    .filter(artist => artist.isActive === '1')
    .map((artist) => {
```

When hitting the API directly in the browser, I get the same error I'm seeing from the `fetch`
**https://www.analogstudios.net/api/artists**
Screen Shot 2023-12-04 at 12 21 28

File: packages/cli/src/lifecycles/bundle.js [hunk 1]
```diff
@@ -163,11 +163,13 @@ async function bundleStyleResources(compilation, resourcePlugins) {
...
 async function bundleApiRoutes(compilation) {
   // https://rollupjs.org/guide/en/#differences-to-the-javascript-api
-  const [rollupConfig] = await getRollupConfigForApis(compilation);
+  const apiConfigs = await getRollupConfigForApis(compilation);
 
-  if (rollupConfig.input.length !== 0) {
-    const bundle = await rollup(rollupConfig);
-    await bundle.write(rollupConfig.output);
+  if (apiConfigs.length > 0 && apiConfigs[0].input.length !== 0) {
+    apiConfigs.forEach(async rollupConfig => {
```
Comment (thescientist13) [inline packages/cli/src/lifecycles/bundle.js:169]: If we want to do something `async` in a loop, we should use a `for ... of` or `for ... in` instead.   `forEech` is synchronous and combining these two can lead to unexpected behaviors (something I would like to more thoroughly clean up in #823 )
```js
for (const rollupConfig of apiConfigs) {
  const bundle = await rollup(rollupConfig);
  await bundle.write(rollupConfig.output);
}
```

**Comment by thescientist13** [comment]
> As I work through the errors I'm seeing things like in-use ports, missing file imports, etc. I'm just not sure whether what I'm seeing is environmental or code-based due to my changes.

Could you be running other Greenwood project's at the same time by chance, occupying ports `1984` or `8080`?  Also, depending on the process exits when running tests, it might not teardown the server running as part of a test case, thus causing a zombie process.

One way to confirm if a process is running is run `yarn develop` or `yarn serve` in this project and see if either of them crash.  You can also do a `ps aux | grep node` to find running Node instances and see if `develop` or `serve` commands are running.  You can then kill the process using `kill -9 <PID>`

Another tip is to pass the [debug flag to `new Runner`](https://github.com/ProjectEvergreen/greenwood/blob/master/.github/CONTRIBUTING.md#running-tests) to get all console output from Greenwood
```js
runner = new Runner(true);
```

Fwiw, I ran all the tests locally from this branch and only got 8 tests failing, exclusively related to API behaviors, which is also what [CI is reporting](https://github.com/ProjectEvergreen/greenwood/actions/runs/7117561854/job/19378560749), so seems like your blast radius is small 💣 😃 
<details>
  <pre>
    3070 passing (5m)
  30 pending
  8 failing

  1) Serve Greenwood With:
       API Routes
         Serve command with API specific behaviors for an HTML ("fragment") API
           should return a 200 status:

      AssertionError: expected 500 to equal 200
      + expected - actual

      -500
      +200

      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/test/cases/serve.default.api/serve.default.api.spec.js:109:36)
      at process.processImmediate (node:internal/timers:471:21)

  2) Serve Greenwood With:
       API Routes
         Serve command with API specific behaviors for an HTML ("fragment") API
           should return a custom status message:

      AssertionError: expected 'Internal Server Error' to equal 'SUCCESS!!!'
      + expected - actual

      -Internal Server Error
      +SUCCESS!!!

      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/test/cases/serve.default.api/serve.default.api.spec.js:114:40)
      at process.processImmediate (node:internal/timers:471:21)

  3) Serve Greenwood With:
       API Routes
         Serve command with API specific behaviors for an HTML ("fragment") API
           should return the correct content type:

      AssertionError: expected 'text/plain; charset=utf-8' to equal 'text/html'
      + expected - actual

      -text/plain; charset=utf-8
      +text/html

      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/test/cases/serve.default.api/serve.default.api.spec.js:119:57)
      at process.processImmediate (node:internal/timers:471:21)

  4) Serve Greenwood With:
       API Routes
         Serve command with API specific behaviors for an HTML ("fragment") API
           should return the correct response body:
     AssertionError: expected 'Internal Server Error' to include '<h1>Hello Greenwood!!!</h1>'
      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/test/cases/serve.default.api/serve.default.api.spec.js:124:25)
      at process.processImmediate (node:internal/timers:471:21)

  5) Serve Greenwood With:
       A Server Rendered Application (SSR) with API Routes importing TypeScript
         Serve command with API specific behaviors for an HTML ("fragment") API
           should return a 200 status:

      AssertionError: expected 500 to equal 200
      + expected - actual

      -500
      +200

      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/plugin-typescript/test/cases/exp-serve.ssr/exp-serve.ssr.spec.js:75:36)
      at process.processImmediate (node:internal/timers:471:21)

  6) Serve Greenwood With:
       A Server Rendered Application (SSR) with API Routes importing TypeScript
         Serve command with API specific behaviors for an HTML ("fragment") API
           should return a custom status message:

      AssertionError: expected 'Internal Server Error' to equal 'OK'
      + expected - actual

      -Internal Server Error
      +OK

      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/plugin-typescript/test/cases/exp-serve.ssr/exp-serve.ssr.spec.js:79:40)
      at process.processImmediate (node:internal/timers:471:21)

  7) Serve Greenwood With:
       A Server Rendered Application (SSR) with API Routes importing TypeScript
         Serve command with API specific behaviors for an HTML ("fragment") API
           should return the correct content type:

      AssertionError: expected 'text/plain; charset=utf-8' to equal 'text/html'
      + expected - actual

      -text/plain; charset=utf-8
      +text/html

      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/plugin-typescript/test/cases/exp-serve.ssr/exp-serve.ssr.spec.js:83:57)
      at process.processImmediate (node:internal/timers:471:21)

  8) Serve Greenwood With:
       A Server Rendered Application (SSR) with API Routes importing TypeScript
         Serve command with API specific behaviors for an HTML ("fragment") API
           should make sure to have the expected CSS inlined into the page for each <app-card>:

      AssertionError: expected 0 to equal 2
      + expected - actual

      -0
      +2

      at Context.<anonymous> (file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/plugin-typescript/test/cases/exp-serve.ssr/exp-serve.ssr.spec.js:89:42)
      at process.processImmediate (node:internal/timers:471:21)



error Command failed with exit code 8.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
  </pre>
<details>

----

> Also, why type of tests are running with test:exp? They seem to be E2E since the CLI is multi-layered, but, are the test cases independent/standalone, or do they rely on suites running in sequence?

All test cases are pretty much the same, they run the CLI and validate some output, so in the context of a CLI app I suppose they would be E2E tests?  (technically though for something like the static router feature, I would love to run it in a browser).  But no, the answer is more boring than that.

The `test:exp` is just a way to distinguish tests that need to run with the `--experimental-loader` flag for NodeJS set, that enables Greenwood to support [Custom Imports](https://www.greenwoodjs.io/docs/server-rendering/#custom-imports).   You will see these test cases prefixed with exp in their folder name.

> Is using `.only` a recommended approach for this? (I thought that's what you were doing during the walkthrough).

Not related to `test:exp`, it's a just a convenient way to "solo" / isolate instead of running all the specs.  Often I'm just iterating TDD style with one test case file until I get it right, then I run against the rest and hope for the best.  😅 

But then if others fail, I will `.only` my way towards solving them all.

----

> The latest suite I'm chasing is the `serve.default.api.spec.js` suite. The `api/fragment.js` is failing to be created, but the other API files do get created for those tests.  That sounds like it could be bundling related for sure, but since other files are being generated, maybe not?

I wonder if it could be related to https://github.com/ProjectEvergreen/greenwood/pull/1186#discussion_r1418257440?  I confirmed that rollup config had an entry when logging `input` but yeah, no output file.  I would see if refactoring those for loops helps and I can take a look into it as well, since the code seems pretty on the 💰 from what I can tell.

**Comment by DevLab2425** [comment]
I have passing tests locally, but the CI is still failing many times. I'll use the CI reports to try to track things down, but having difficulty identifying things without failures locally.

**Review by thescientist13 — COMMENTED** [review]
Hmm, not sure why GitHub Actions is failing so hard here, even after re-running the jobs, but I ran `yarn test:exp` locally and all the specs passed.  😕 

While i look into that, I think last steps would be to step into each of the adapter plugins and see if we can now remove all the extraneous file copying we had to do.
- Vercel - https://github.com/ProjectEvergreen/greenwood/blob/master/packages/plugin-adapter-vercel/src/index.js#L107
- Netlify - https://github.com/ProjectEvergreen/greenwood/blob/master/packages/plugin-adapter-netlify/src/index.js#L118

Thanks for all this, please keep the faith!  😇 

----

I'm going to play around with this locally a bit to examine the new output and will aim to patch these changes into one of the respective demo repos to do some final validation testing, which I think I can do this weekend.  (and also the test cases)

File: packages/cli/src/lifecycles/bundle.js [hunk 2]
```diff
@@ -163,11 +163,15 @@ async function bundleStyleResources(compilation, resourcePlugins) {
 
 async function bundleApiRoutes(compilation) {
   // https://rollupjs.org/guide/en/#differences-to-the-javascript-api
-  const [rollupConfig] = await getRollupConfigForApis(compilation);
+  const apiConfigs = await getRollupConfigForApis(compilation);
 
-  if (rollupConfig.input.length !== 0) {
-    const bundle = await rollup(rollupConfig);
-    await bundle.write(rollupConfig.output);
+  if (apiConfigs.length > 0 && apiConfigs[0].input.length !== 0) {
```
Comment (thescientist13) [inline packages/cli/src/lifecycles/bundle.js:168]: minor nit: but for each of these loops, if we use `for ... of` we can get access to the index directly.
```js
async function bundleApiRoutes(compilation) {
  // https://rollupjs.org/guide/en/#differences-to-the-javascript-api
  const apiConfigs = await getRollupConfigForApis(compilation);

  if (apiConfigs.length > 0 && apiConfigs[0].input.length !== 0) {
    for (const config of apiConfigs) {
      const bundle = await rollup(config);

      await bundle.write(config.output);
    }
  }
}
```

**Comment by thescientist13** [comment]
@DevLab2425 
OK, so I think I found the culprit.  Looks like Greenwood fails on Node 18.19.0, which is the version I see running in GitHub Actions.  If I `nvm use 18.19.0` on my machine, I can reproduce the error in CI and with some additional logging can get some helpful information.  I basically see a lot of errors like this
```js
  TypeError [ERR_INVALID_URL]: Invalid URL
      at new NodeError (node:internal/errors:405:5)
      at new URL (node:internal/url:676:13)
      at file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/src/lifecycles/serve.js:129:17
      at dispatch (/Users/owenbuckley/Workspace/project-evergreen/greenwood/node_modules/koa-compose/index.js:42:32)
      at file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/src/lifecycles/serve.js:123:11
      at dispatch (/Users/owenbuckley/Workspace/project-evergreen/greenwood/node_modules/koa-compose/index.js:42:32)
      at file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/src/lifecycles/serve.js:87:11
      at dispatch (/Users/owenbuckley/Workspace/project-evergreen/greenwood/node_modules/koa-compose/index.js:42:32)
      at file:///Users/owenbuckley/Workspace/project-evergreen/greenwood/packages/cli/src/lifecycles/serve.js:54:11
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Taking a look, maybe it's something was doing wrong, or perhaps a regression in Node.  What version are you using?


Will report back!  🫡

**Review by thescientist13 — COMMENTED** [review]
Hey @DevLab2425 , just have some updates to share:

Regarding the build failures, I've opened #1191 to remove the offending code path completely until I can properly fix / implement it later in https://github.com/ProjectEvergreen/greenwood/issues/684.  For local dev, I would just use Node < 18..19.0

I'll plan to get it merged Wednesday evening after the weekly meeting.

----

I also got around to doing some testing over the weekend of these changes into our [Vercel demo repo](https://github.com/ProjectEvergreen/greenwood-demo-adapter-vercel/pull/22) and it worked great!  I even patched in some cleanup of the Vercel adapter to remove a few of the `cp` calls we were making and was able to see a definite reduction in file generated.  💪 

```sh
# before CLI
➜  greenwood-demo-adapter-vercel git:(main) tree public
public
├── 1733327959.ed8fa148.js.map
├── 396762157.af333475.js.map
├── 404.html
├── 758806547.5433709a.js.map
├── 949355521.f4273cc9.js.map
├── __products.js
├── _products.js
├── api
│   ├── assets
│   ├── card.51aef4b7.js
│   ├── fragment.js
│   ├── greeting.js
│   ├── products.c34c47ae.js
│   └── search.js
├── card.51aef4b7.js
├── card.51aef4b7.js.map
├── favicon.ico
├── graph.json
├── index.html
├── manifest.json
├── resources.json
├── search
│   └── index.html
└── styles
    └── main.19738545.css

5 directories, 21 files
```

```sh
# after CLI
➜  greenwood-demo-adapter-vercel git:(chore/upgrade-greenwood-0.30.0) ✗ tree public
public
├── 1733327959.ed8fa148.js.map
├── 396762157.af333475.js.map
├── 404.html
├── 758806547.5433709a.js.map
├── 949355521.f4273cc9.js.map
├── __products.js
├── _products.js
├── api
│   ├── assets
│   ├── card.51aef4b7.js
│   ├── fragment.js
│   ├── greeting.js
│   └── search.js
├── card.51aef4b7.js
├── card.51aef4b7.js.map
├── favicon.ico
├── graph.json
├── index.html
├── manifest.json
├── resources.json
├── search
│   └── index.html
└── styles
    └── main.19738545.css

5 directories, 20 files
```


```sh
# before adapter
➜  greenwood-demo-adapter-vercel git:(89730d7) ✗ tree .vercel
.vercel
└── output
    ├── config.json
    ├── functions
    │   ├── api
    │   │   ├── fragment.func
    │   │   │   ├── assets
    │   │   │   ├── card.51aef4b7.js
    │   │   │   ├── fragment.js
    │   │   │   ├── index.js
    │   │   │   ├── package.json
    │   │   │   └── products.c34c47ae.js
    │   │   ├── greeting.func
    │   │   │   ├── assets
    │   │   │   ├── card.51aef4b7.js
    │   │   │   ├── greeting.js
    │   │   │   ├── index.js
    │   │   │   ├── package.json
    │   │   │   └── products.c34c47ae.js
    │   │   └── search.func
    │   │       ├── assets
    │   │       ├── card.51aef4b7.js
    │   │       ├── index.js
    │   │       ├── package.json
    │   │       ├── products.c34c47ae.js
    │   │       └── search.js
    │   └── products.func
    │       ├── __products.js
    │       ├── _products.js
    │       ├── card.51aef4b7.js
    │       ├── index.js
    │       └── package.json
    └── static
        ├── 1733327959.ed8fa148.js.map
        ├── 396762157.af333475.js.map
        ├── 404.html
        ├── 758806547.5433709a.js.map
        ├── 949355521.f4273cc9.js.map
        ├── __products.js
        ├── _products.js
        ├── api
        │   ├── assets
        │   ├── card.51aef4b7.js
        │   ├── fragment.js
        │   ├── greeting.js
        │   ├── products.c34c47ae.js
        │   └── search.js
        ├── card.51aef4b7.js
        ├── card.51aef4b7.js.map
        ├── favicon.ico
        ├── graph.json
        ├── index.html
        ├── manifest.json
        ├── resources.json
        ├── search
        │   └── index.html
        └── styles
            └── main.19738545.css

16 directories, 42 files
```

```sh
# after adapter
➜  greenwood-demo-adapter-vercel git:(chore/upgrade-greenwood-0.30.0) ✗ tree .vercel
.vercel
└── output
    ├── config.json
    ├── functions
    │   ├── api
    │   │   ├── fragment.func
    │   │   │   ├── card.51aef4b7.js
    │   │   │   ├── fragment.js
    │   │   │   ├── index.js
    │   │   │   └── package.json
    │   │   ├── greeting.func
    │   │   │   ├── card.51aef4b7.js
    │   │   │   ├── greeting.js
    │   │   │   ├── index.js
    │   │   │   └── package.json
    │   │   └── search.func
    │   │       ├── card.51aef4b7.js
    │   │       ├── index.js
    │   │       ├── package.json
    │   │       └── search.js
    │   └── products.func
    │       ├── __products.js
    │       ├── _products.js
    │       ├── index.js
    │       └── package.json
    └── static
        ├── 1733327959.ed8fa148.js.map
        ├── 396762157.af333475.js.map
        ├── 404.html
        ├── 758806547.5433709a.js.map
        ├── 949355521.f4273cc9.js.map
        ├── __products.js
        ├── _products.js
        ├── api
        │   ├── assets
        │   ├── card.51aef4b7.js
        │   ├── fragment.js
        │   ├── greeting.js
        │   └── search.js
        ├── card.51aef4b7.js
        ├── card.51aef4b7.js.map
        ├── favicon.ico
        ├── graph.json
        ├── index.html
        ├── manifest.json
        ├── resources.json
        ├── search
        │   └── index.html
        └── styles
            └── main.19738545.css

13 directories, 37 files
```


One call out is that when using `new URL` in a file, a chunk will always be generated as opposed to being "inlined" when it is an `import` since that's just how that works.  As you can see in the output above, I still had to copy _card.js_ into every API function folder, even though it is technically only needed for two of the APIs.

It would be interesting to see if there's anyway to link these to the entry point that produced it, so we could have a stronger coupling between assets of an API or SSR page.  Probably something I can make another ticket for.

----

After Adapter (Final)
```sh
➜  greenwood-demo-adapter-vercel git:(chore/upgrade-greenwood-0.30.0) ✗ tree .vercel 
.vercel
└── output
    ├── config.json
    ├── functions
    │   ├── api
    │   │   ├── fragment.func
    │   │   │   ├── card.51aef4b7.js
    │   │   │   ├── fragment.js
    │   │   │   ├── index.js
    │   │   │   └── package.json
    │   │   ├── greeting.func
    │   │   │   ├── greeting.js
    │   │   │   ├── index.js
    │   │   │   └── package.json
    │   │   └── search.func
    │   │       ├── card.51aef4b7.js
    │   │       ├── index.js
    │   │       ├── package.json
    │   │       └── search.js
    │   └── products.func
    │       ├── index.js
    │       ├── package.json
    │       ├── products.route.chunk.d46126ed.js
    │       └── products.route.js
    └── static
        ├── 1733327959.ed8fa148.js.map
        ├── 396762157.af333475.js.map
        ├── 404.html
        ├── 758806547.5433709a.js.map
        ├── 949355521.f4273cc9.js.map
        ├── api
        │   ├── card.51aef4b7.js
        │   ├── fragment.js
        │   ├── greeting.js
        │   └── search.js
        ├── card.51aef4b7.js
        ├── card.51aef4b7.js.map
        ├── favicon.ico
        ├── graph.json
        ├── index.html
        ├── manifest.json
        ├── products.route.chunk.d46126ed.js
        ├── products.route.js
        ├── resources.json
        ├── search
        │   └── index.html
        └── styles
            └── main.19738545.css

12 directories, 36 files
```

**Comment by thescientist13** [comment]
OK, #1190 has been merged 👍

**Review by thescientist13 — COMMENTED** [review]
Sweet, thanks for getting the ball rolling on this.  Was able to explore some refactoring opportunities here to remove some very obtuse code, so that felt really good.  In that refactoring found (but not 100 known why) a bug in WCC, so just patched that in for now.

Just going to add some docs on the bundle output changes and tinker with final bundle mapping idea, but this should be good to land as soon as we release the next version of WCC.  🙌

**Review by thescientist13 — COMMENTED** [review]
Sweet, felt to clean up some things here!

Just need to publish some upstream changes for WCC then this will be good to go!

**Pull request merged**